test-it/karma/publicapi/TypeReflection.Test.js

```diff
@@ -0,0 +1,140 @@
+    var fixtures = window.testHelpers.fixtures;
+
+    var vireo = new Vireo();
+
+    var publicApiMultipleTypesViaUrl = fixtures.convertToAbsoluteFromFixturesDir('publicapi/MultipleTypes.via');
+    var viName = 'MyVI';
+
+    var typeDescriptor = (function () {
+        var returnTypeName = function (name) {
+            var typeName = name;
+            return function () {
+                return typeName;
+            };
+        };
+
+        return {
+            visitBoolean: returnTypeName('Boolean'),
+            visitEnum: returnTypeName('Enum'),
+            visitInt8: returnTypeName('Int8'),
+            visitInt16: returnTypeName('Int16'),
+            visitInt32: returnTypeName('Int32'),
+            visitInt64: returnTypeName('Int64'),
+            visitUInt8: returnTypeName('UInt8'),
+            visitUInt16: returnTypeName('UInt16'),
+            visitUInt32: returnTypeName('UInt32'),
+            visitUInt64: returnTypeName('UInt64'),
+            visitSingle: returnTypeName('Single'),
+            visitDouble: returnTypeName('Double'),
+            visitString: returnTypeName('String'),
+            visitComplexSingle: returnTypeName('ComplexSingle'),
+            visitComplexDouble: returnTypeName('ComplexDouble'),
+            visitAnalogWaveform: returnTypeName('AnalogWaveform'),
+            visitTimestamp: returnTypeName('Timestamp'),
+            visitPath: returnTypeName('Path'),
+            visitArray: returnTypeName('Array'),
+            visitCluster: returnTypeName('Cluster')
+        };
+    }());
+
+    beforeAll(function (done) {
+        fixtures.preloadAbsoluteUrls([
+            publicApiMultipleTypesViaUrl
+        ], done);
+    });
+
+    beforeAll(function () {
+        vireoRunner.rebootAndLoadVia(vireo, publicApiMultipleTypesViaUrl);
+    });
+
+    var getTypeName = function (path) {
+        return vireo.eggShell.reflectOnValueRef(typeDescriptor, vireo.eggShell.findValueRef(viName, path));
+    };
+
+    var reflectOnEmptyVisitor = function (path) {
+        var thePath = path;
+        return function () {
+            vireo.eggShell.reflectOnValueRef({}, vireo.eggShell.findValueRef(viName, thePath));
+        };
+    };
+
+    describe('reflectOnValueRef', function () {
+        describe('error handling', function () {
+            it('throws when typeDescriptor is not an object', function () {
+                var throwingFunction = function () {
+                    vireo.eggShell.reflectOnValueRef([]);
+                };
+
+                expect(throwingFunction).toThrow();
+            });
+
+            it('throws when valueRef is not an object', function () {
+                var throwingFunction = function () {
+                    vireo.eggShell.reflectOnValueRef({}, []);
+                };
+
+                expect(throwingFunction).toThrow();
+            });
+
+            it('throws when visitor does not have visit methods for all types', function () {
```

Comment:
Generally if the use case isn't well know I think it's better to throw and then make it less strict if needed. It's a lot trickier to start throwing exceptions than to stop (assuming people are not using exceptions for flow control :P)

Reply:
Any use-case you have in mind?

Reply:
I like the proposed implementation because if a user decides to write a partial visitor and vireo reflects on a type that the user didn't specify it will throw. The alternative by not throwing would be to give the user the responsibility of keeping track if any of their functions in their visitor was called. Now that I've written down the alternative I'm more convinced to leave it as it is :P

Reply:
Yes, please throw, and let the user implement a visitor function that does nothing if they don't want to deal with the type.